### PR TITLE
Np 45

### PR DESCRIPTION
**Pull request by lcaloguerea**
[NP-45](https://pappcorn.atlassian.net/secure/RapidBoard.jspa?rapidView=112&projectKey=NP&selectedIssue=NP-45)
[NP-46](https://pappcorn.atlassian.net/secure/RapidBoard.jspa?rapidView=112&projectKey=NP&modal=detail&selectedIssue=NP-46)

# Change description
- Added Dynamically generated codes for schools
- Added School creator
- Added geo localization
- Added School service
- Added School Model

# Evidence
https://user-images.githubusercontent.com/7730194/107514731-cfbe1f80-6b88-11eb-9529-d400225fb3ac.mp4

### CODE REVIEW

File: apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts [hunk 1]
```diff
@@ -1,8 +1,211 @@
-import { Component } from '@angular/core';
+/// <reference types="@types/googlemaps" />
+import { Component, ElementRef, NgZone, OnInit, ViewChild } from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { MapsAPILoader, MouseEvent } from '@agm/core';
+import { GeneralService } from '../../services/generalService/general.service';
+import { School } from '../../../../../../utils/models/school';
```
Comment (ruedagato) [inline apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts:6]: use @pang/interface

File: apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts [hunk 2]
```diff
@@ -1,8 +1,211 @@
...
+    private generalService: GeneralService,
+    private schoolsService: SchoolService,
+  ) {
+    this.dataSource = new MatTableDataSource();
+  }
+
+  ngOnInit() {
+    this.isLoading = true;
+    //subscribe to schools to load data table
+    this.schoolSubscription = this.schoolsService.getAll().subscribe((schools) => {
```
Comment (ruedagato) [inline apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts:86]: maybe you can move all map login to a service to be more scalable

File: apps/dashboard/src/app/pages/login/login.component.ts [hunk 3]
```diff
@@ -22,17 +23,21 @@ export class LoginComponent implements OnInit {
...
     const { password } = this.loginForm.value;
     if ((email && email != 'admin@newpangea.com') || (password && password != '12345678')) {
       alert('Incorrect credentials');
     } else {
-      console.log('login in');
-      console.log(this.router.navigate(['/', 'dashboard']));
-      this.router.navigate(['/', 'dashboard']);
+      this.isLogin = true;
+      this.loginService.loginWithEmail(email, password).then((fireUser) => {
+        if (fireUser) {
```
Comment (ruedagato) [inline apps/dashboard/src/app/pages/login/login.component.ts:33]: check logic for invalid credentials

File: apps/dashboard/src/app/shared/services/generalService/general.service.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import { AngularFirestore } from '@angular/fire/firestore';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class GeneralService {
+  constructor(private db: AngularFirestore) {}
+
+  getFirestoreId(): string {
+    return this.db.createId();
```
Comment (ruedagato) [inline apps/dashboard/src/app/shared/services/generalService/general.service.ts:11]: not is necessary to create un service to do this, AngularFirestore already is a service that does it

File: apps/dashboard/src/app/shared/services/schoolService/school.service.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { Injectable, EventEmitter } from '@angular/core';
+import { AngularFireAuth } from '@angular/fire/auth';
+import { AngularFirestore } from '@angular/fire/firestore';
+import { School } from '../../../../../../utils/models/school';
```
Comment (ruedagato) [inline apps/dashboard/src/app/shared/services/schoolService/school.service.ts:4]: move to a @pang/interface or create @pang/models

File: apps/dashboard/src/app/shared/services/schoolService/school.service.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class SchoolService {
+  public didLogout: EventEmitter<void> = new EventEmitter();
+
+  constructor(private auth: AngularFireAuth, private db: AngularFirestore) {}
+
+  add(school: School) {
+    return this.db.collection(School.REF).doc(school.key).set(Object.assign({}, school));
```
Comment (ruedagato) [inline apps/dashboard/src/app/shared/services/schoolService/school.service.ts:15]:
```suggestion
    return this.db.collection(School.REF).doc(school.key).set({...school});
```

File: apps/utils/models/school.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+export class School {
```
Comment (ruedagato) [inline apps/utils/models/school.ts:1]: utils folder inside app is a bad practice, move it to lib folder and you can guide yourself like constants and interfaces folders

File: apps/utils/models/user.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  cellphone: string;
+  gender: number;
+  studyLevel: number;
+  constructor(key: string, email: string) {
+    super(key, email);
+    this.role = ROLE_TYPES.ADMIN;
+  }
+}
+
+export enum ROLE_TYPES {
```
Comment (ruedagato) [inline apps/utils/models/user.ts:25]: move to constants folder

File: libs/styles/global.scss [hunk 9]
```diff
@@ -7,7 +7,7 @@
 html,
 body {
   margin: 0;
-  font-family: 'Inter', sans-serif;
+  font-family: 'Inter', sans-serif !important;
```
Comment (ruedagato) [inline libs/styles/global.scss:10]: you must avoid important

**Review by ruedagato — APPROVED** [review]
Check all comments to fix them in a new pr

File: apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts (hunk 1 above)
Comment (lcaloguerea) [reply apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts:6]: Solved

File: apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts (hunk 2 above)
Comment (lcaloguerea) [reply apps/dashboard/src/app/shared/components/sidebar-menu/sidebar-menu.component.ts:86]: Solved

**Pull request merged**